### DISCUSSION TITLE
data dependency possibilities with CUE

### DISCUSSION THREAD

**Discussion by vgeordy** [cue-lang/cue#2879] [Q&A] [answered]
Hi, I am completely new to CUE and have an inquiry about data dependency possibilities.

lets say that I have a data-dependency.cue file that is defined like this: 

```
#name: "Alice" | "Bob"
#place("Alice"): "New York" | "London"
#place("Bob"): "Chicago" | "Tokyo"
```
note: I know this is not valid syntax, but this is essentially what I want. The place should depend on the name given.

and a schema.cue that is defined something like this:
```
#Sample &{
    name: #name
    place: #place
}

```
and a data.yaml which we will test for validation based on the files above

```
...
name: Alice
place: "Chicago"
...
```
this validation should fail because Alice doesn't have Chicago as one of her #place. How would I do something like this in cue or go. Note that data-dependency.cue values can change, but #name and #place remain the same and it is something the user will manually define. I understand there are conditional fields but does it work with dynamic data-dependency files?

Thank you for your help!

**Comment by jpluscplusm** [comment, marked as answer]
There are a few different ways to achieve this! Here's a playground link that combines 2 methods in 1 evaluation: https://alpha.cuelang.org/play/?id=tIne24zL-Lv#w=function&i=cue&f=eval&o=cue. The method you use (and there are definitely more methods!) is down to your preference.

Notice that on line 23, both `#Sample1` and `#Sample2` are unified, which is only possible because they *can* unify successfully, and they both represent the same constraints. You shouldn't use both methods simultaneously - I'm just demonstrating that both achieve the same outcome.

The playground isn't showing the errors with Bob's failing example. I believe this is a CUE Playground artifact - if you were to comment out Alice's failing example, the Bob failure would still strike.

**Reply by vgeordy** [reply]
I believe Sample#2 is just what I need. We don't know the values for #name before hand so I am not sure if the conditionals in Sample#1 would work.

Thank you for taking the time to answer my question. This was really helpful!

**Reply by jpluscplusm** [reply]
Glad you got the answer you need 😀

One question, though:

> We don't know the values for #name before hand so I am not sure if the conditionals in Sample#1 would work.

If you don't know the names "beforehand", how will you define the map that I called `#personPlace`?

The lookup map is no different from the conditional approach: you still need to have _some_ way of assigning valid options to people's names. 

CUE lets you separate these different facets into -- say -- different files, as it unifies everything as part of its evaluation process. So you've got the same flexibility across both the methods I suggested: you'll need to have /some/ CUE that references specific names and maps them to cities. 

It /is/ possible to delegate the data side of the people-to-city map to a data format (like YAML), and construct the map dynamically /from/ the data - but that just pushes the mapping definition elsewhere ... it doesn't remove the need for the map entirely.

**Reply by vgeordy** [reply]
sorry, I worded that wrongly the values for #name are known. I was more referring to what you are describing in your last two paragraphs. That is good information to know. Thank you!
